renderdoc/driver/vulkan/vk_acceleration_structure.cpp

```diff
@@ -37,13 +37,410 @@ constexpr VkDeviceSize handleCountSize = 8;
+        geoData.memSize = readbackmem.size;
+        geoData.memAlignment = mrq.alignment;
+
+        geoData.aabbs.stride = aabbInfo.stride;
+
+        metadata->geometryData.push_back(geoData);
+
+        VkAccelerationStructureBuildRangeInfoKHR buildData = rangeInfo;
+        buildData.primitiveOffset = 0;
+
+        metadata->buildRangeInfo.push_back(buildData);
+
+        break;
+      }
+      case VK_GEOMETRY_TYPE_INSTANCES_KHR:
+      {
+        const VkAccelerationStructureGeometryInstancesDataKHR &instanceInfo =
+            geometry.geometry.instances;
+
+        // Find the associated VkBuffer
+        BufferData data = GetDeviceAddressData(instanceInfo.data.deviceAddress);
+        if(!data)
+        {
+          RDCERR("Unable to find VkBuffer for instance data at 0x%llx",
+                 instanceInfo.data.deviceAddress);
+          continue;
+        }
+
+        data.size = rangeInfo.primitiveCount * sizeof(VkAccelerationStructureInstanceKHR);
+
+        // Get the alignment
+        VkMemoryRequirements mrq = {};
+        ObjDisp(device)->GetBufferMemoryRequirements(Unwrap(device), data.buf, &mrq);
+
+        // Allocate copy buffer
+        readbackmem = CreateReadBackMemory(device, data.size, mrq.alignment);
+        if(readbackmem.mem == VK_NULL_HANDLE)
+        {
+          RDCERR("Unable to allocate AS instance input buffer readback memory (size: %u bytes)",
+                 data.size);
+          continue;
+        }
+
+        // Insert copy commands
+        VkBufferCopy region = {
+            data.rao.offset + rangeInfo.primitiveOffset,
+            0,
+            data.size,
+        };
+        ObjDisp(device)->CmdCopyBuffer(Unwrap(commandBuffer), data.buf, readbackmem.buf, 1, &region);
+
+        // Prepare barrier
+        barriers = {data.Barrier()};
+
+        // Store the metadata
+        VkAccelerationStructureInfo::GeometryData geoData;
+        geoData.geometryType = geometry.geometryType;
+        geoData.flags = geometry.flags;
+        geoData.readbackMem = readbackmem.mem;
+        geoData.memSize = readbackmem.size;
+        geoData.memAlignment = mrq.alignment;
+
+        metadata->geometryData.push_back(geoData);
+
+        VkAccelerationStructureBuildRangeInfoKHR buildData = rangeInfo;
+        buildData.primitiveOffset = 0;
+
+        metadata->buildRangeInfo.push_back(buildData);
+
+        break;
+      }
+      default: RDCERR("Unhandled geometry type: %d", geometry.geometryType); return;
+    }
+
+    // Insert barriers to block any other commands until the buffers are copied
+    if(!barriers.empty())
+    {
+      ObjDisp(device)->CmdPipelineBarrier(Unwrap(commandBuffer), VK_PIPELINE_STAGE_TRANSFER_BIT,
+                                          VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, 0, 0, VK_NULL_HANDLE,
+                                          static_cast<uint32_t>(barriers.size()), barriers.data(),
+                                          0, VK_NULL_HANDLE);
+    }
```

Comment:
I think rather than doing individual buffer barriers it would be easier to just do one global memory barrier. I'm not sure the exact difference for drivers but I believe few even use buffer barriers as anything specific, and if any do as soon as we have more than a couple of buffers here it's likely going to decay.

Also note that this barrier after the copies is to ensure that the input buffers aren't overwritten by any _later_ commands, so we should have both source and destination access masks to ensure that.

Reply:
Is there any existing functionality that implements this?  Rather than me re-invent the wheel.

Reply:
It could be the same code as you have here but just one barrier at the end instead of N.